### PR TITLE
ColoredConsoleTarget - Allow using Condition-logic for word highlighting rules

### PR DESCRIPTION
**Pull request by snakefoot**
Condition-Logic is also faster than RegEx. Especially when combined with #3894

### CODE REVIEW

**Comment by snakefoot** [comment]
Before tuning:

image

After tuning (Faster handling of newlines and reduced number of console color changes):

image

**Comment by 304NotModified** [comment]
Cool!

Would be nice if we could cover these lines:

image


image

**Comment by snakefoot** [comment]
> Would be nice if we could cover these lines:

Yes I'm missing unit-tests. Just playing around with comparing Microsoft AddConsole and NLog ColoredConsole. Noticed that Microsoft AddConsole is not cheap.

**Comment by snakefoot** [comment]
@304NotModified Was just in a battle with code-smells, and needed some retries before it worked. Also the Azure-Build-Server sometimes like to timeout, so one needs an extra run.

**Comment by 304NotModified** [comment]
Well it's stuck on azure then,as the timeout is 60min and it should be < 20min.

File: src/NLog/Targets/ColoredConsoleAnsiPrinter.cs [hunk 1]
```diff
@@ -51,7 +51,7 @@ public TextWriter AcquireTextWriter(TextWriter consoleStream, StringBuilder reus
             return new StringWriter(reusableBuilder ?? new StringBuilder(50), consoleStream.FormatProvider);
         }
 
-        public void ReleaseTextWriter(TextWriter consoleWriter, TextWriter consoleStream, ConsoleColor? oldForegroundColor, ConsoleColor? oldBackgroundColor, bool flush)
+        public void ReleaseTextWriter(TextWriter consoleWriter, TextWriter consoleStream, ConsoleColor? orgForegroundColor, ConsoleColor? orgBackgroundColor, bool flush)
```
Comment (304NotModified) [inline src/NLog/Targets/ColoredConsoleAnsiPrinter.cs:54]: please write out org, prev, etc

File: src/NLog/Targets/ConsoleWordHighlightingRule.cs [hunk 2]
```diff
@@ -143,8 +150,13 @@ public bool IgnoreCase
         /// </summary>
         public Regex CompiledRegex => _regexHelper.Regex;
 
-        internal MatchCollection Matches(string message)
+        internal MatchCollection Matches(LogEventInfo logEvent, string message)
         {
+            if (Condition != null && !true.Equals(Condition.Evaluate(logEvent)))
```
Comment (304NotModified) [inline src/NLog/Targets/ConsoleWordHighlightingRule.cs:155]: false.Equals ;)?

Comment (snakefoot) [reply src/NLog/Targets/ConsoleWordHighlightingRule.cs:155]: No opinion. Will fix.

File: src/NLog/Targets/ColoredConsoleAnsiPrinter.cs (hunk 1 above)
Comment (snakefoot) [reply src/NLog/Targets/ColoredConsoleAnsiPrinter.cs:54]: Uhh suddenly old is good again. Will change to old then

**Comment by snakefoot** [comment]
Sigh. Now my refactoring revealed that we never really tested other control-characters than newline. New unit-test is needed.

**Comment by 304NotModified** [comment]
> Sigh. Now my refactoring revealed that we never really tested other control-characters than newline. New unit-test is needed.

So convert this PR to draft? Or is this fixed in the force push?

**Comment by snakefoot** [comment]
>So convert this PR to draft? Or is this fixed in the force push?

Fixed as code-coverage is back at 100%

**Comment by 304NotModified** [comment]
ow indeed! NICE!

**Comment by 304NotModified** [comment]
Thanks! merged!

**Comment by snakefoot** [comment]
Updated wiki: https://github.com/NLog/NLog/wiki/ColoredConsole-target

**Pull request merged**